datafusion/physical-expr/src/window/window_frame_state.rs

```diff
@@ -0,0 +1,972 @@
+                    if SEARCH_SIDE == is_descending {
+                        // TODO: Handle positive overflows
+                        value.add(delta)
+                    } else if value.is_unsigned() && value < delta {
+                        // NOTE: This gets a polymorphic zero without having long coercion code for ScalarValue.
+                        //       If we decide to implement a "default" construction mechanism for ScalarValue,
+                        //       change the following statement to use that.
+                        value.sub(value)
+                    } else {
+                        // TODO: Handle negative overflows
+                        value.sub(delta)
+                    }
+                })
+                .collect::<Result<Vec<ScalarValue>>>()?
+        } else {
+            current_row_values
+        };
+        // `BISECT_SIDE` true means bisect_left, false means bisect_right
+        bisect::<BISECT_SIDE>(range_columns, &end_range, sort_options)
+    }
+}
+
+// In GROUPS mode, rows with duplicate sorting values are grouped together.
+// Therefore, there must be an ORDER BY clause in the window definition to use GROUPS mode.
+// The syntax is as follows:
+//     GROUPS frame_start [ frame_exclusion ]
+//     GROUPS BETWEEN frame_start AND frame_end [ frame_exclusion ]
+// The optional frame_exclusion specifier is not yet supported.
+// The frame_start and frame_end parameters allow us to specify which rows the window
+// frame starts and ends with. They accept the following values:
+//    - UNBOUNDED PRECEDING: Start with the first row of the partition. Possible only in frame_start.
+//    - offset PRECEDING: When used in frame_start, it refers to the first row of the group
+//                        that comes "offset" groups before the current group (i.e. the group
+//                        containing the current row). When used in frame_end, it refers to the
+//                        last row of the group that comes "offset" groups before the current group.
+//    - CURRENT ROW: When used in frame_start, it refers to the first row of the group containing
+//                   the current row. When used in frame_end, it refers to the last row of the group
+//                   containing the current row.
+//    - offset FOLLOWING: When used in frame_start, it refers to the first row of the group
+//                        that comes "offset" groups after the current group (i.e. the group
+//                        containing the current row). When used in frame_end, it refers to the
+//                        last row of the group that comes "offset" groups after the current group.
+//    - UNBOUNDED FOLLOWING: End with the last row of the partition. Possible only in frame_end.
+
+// This structure encapsulates all the state information we require as we
+// scan groups of data while processing window frames.
+#[derive(Debug, Default)]
+pub struct WindowFrameStateGroups {
```

Comment:
This structure is fine -- I just wanted to point out a pattern for writing state machines in Rust is to encapsulate each state in an enum variant, and then you can use the type system to ensure the state transitions are valid

So like

```rust 
pub enum WindowFrameStateGroupsState {
  Start { },
  InGroup { ... }
  End {}
}
```

Or something similar